### ISSUE TITLE
Peer review 2

### ISSUE THREAD

**Issue by Veikkosuhonen** [matiastamsi/tiralabra#2]
# Review

This project and the deciphering seem very interesting, I wish I had the time to figure out how the trie-algorithm works, it looks very intriguing. A user guide would be nice but I think I got the application working well except for the cracking which I see is still a WIP. It seems that the crack-method never reaches the condition where `allCorrect(p)` is true and always returns a null. I hope my inputs weren't wrong or uncrackable. 

I think you've done great job keeping the code simple and clean even when its in active development. I'm only able to very vaguely understand the idea of the cracking algorithm but the code structure was very easy to follow and understandable. It just looks pretty when all methods are around 20 lines long at most! The crack method is pretty long but I get that its for a good reason. The class structure and all of that seems to be well thought out and in a good shape and there's no weird design choices or anything.

The documentation is overall just great and consistent, if you still want to improve something, you could add a bit more comments inside the longer methods for the if-statements and for-loops, that would make them a lot easier for readers to follow.

## Some practical suggestions

These are pretty opinionated and just how I would do things.

Use the String-method toLowerCase on inputs. I noticed that things break when for example the first word I give to the cipher generator is "I", instead of "i". It's also fine using some library methods if it helps with development and you can implement your own later once the core parts of the project are working. 

Setup a streamlined development mode. You probably have some good routine of running the program quickly during development but I found it quite laborious to test, because I have to go through the multiple cli dialogs to get to the cracking part and have to manually type a lot. Besides the "crack" and "exit" options, add a "test" option which tries to crack a predefined cipher. Or write an integration test case for that. Maybe you already have something like this, I just couldn't figure it out.

Use ArrayList and Queue. I see that you've already implemented a queue for the Letter-class so this suggestion is a bit late. It might also be a good idea to separate the queue implementation to its own class. Also in the TrieNode class, you are using an array of children that needs to be resized, so you could use a Java List for that until you implement your own resizable array. Overall I personally would focus on the core algorithm and not worry about implementing all the data structures with custom code (this is very subjective of course).

In conclusion I think overall everything is going very well in you project and I wish you best of luck for the rest of the course :) I definitely want to try this out once its finished.

**Comment by matiastamsi** [comment]
Thank you for your feedback! It was really useful and I will definitely keep those things in my mind.
